### PR TITLE
Status and progress message in UI when a deployment is started or stopped. Auto-updates on completion.

### PR DESCRIPTION
**Pull request by tydangelo18**

### CODE REVIEW

File: components/Firebird/Deployments/ExistingDeployment/SelectedDeployment/StatusStream.tsx [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  if (status === DeploymentStatus.ACTIVE) return null
+
+  const eventDisplayNames =
+    status === DeploymentStatus.STARTING
+      ? {
+          [EventStreamEvents.START_INSTANCE]: "Starting Instance",
+        }
+      : {
+          [EventStreamEvents.STOP_INSTANCE]: "Stopping Instance",
+        }
```
Comment (KwilLuke) [inline components/Firebird/Deployments/ExistingDeployment/SelectedDeployment/StatusStream.tsx:26]: Could you clean this up? Bit hard to read.

Something similar to how Martin organized here would be great: https://github.com/kwilteam/db-admin/blob/53d4bfe9caaad5e1a6c01ee8963b844f13b5a576/components/Firebird/Deployments/ExistingDeployment/SelectedDeployment/DeploymentStatusStream.tsx#L21

File: hooks/firebird/use-event-stream.tsx [hunk 2]
```diff
@@ -105,14 +110,47 @@ const useDeploymentEventStream = (deploymentId: string) => {
...
+      return "follow_deployment_start"
+    } else if (
+      deploymentStatus === DeploymentStatus.STOPPING ||
+      deploymentStatus === DeploymentStatus.STOPPED
+    ) {
+      return "follow_deployment_stop"
+    } else if (deploymentStatus === DeploymentStatus.DEPLOYING) {
+      return "follow_deployment"
+    } else if (deploymentStatus === DeploymentStatus.ACTIVE) {
+      return lastDeploymentStatus === DeploymentStatus.STARTING
```
Comment (KwilLuke) [inline hooks/firebird/use-event-stream.tsx:145]: What is the role of `lastDeploymentStatus`? Having a hard time following the logic here

Comment (KwilLuke) [reply hooks/firebird/use-event-stream.tsx:145]: When does this logic get hit?

**Review by KwilLuke — COMMENTED** [review]
I am still a bit confused on what is going on here... might just be the end of friday feeling

File: hooks/firebird/use-event-stream.tsx [hunk 3]
```diff
@@ -105,14 +110,47 @@ const useDeploymentEventStream = (deploymentId: string) => {
...
+    if (deploymentStatus === DeploymentStatus.STARTING) {
+      setLastDeploymentStatus(DeploymentStatus.STARTING)
+    } else if (
+      deploymentStatus === DeploymentStatus.STOPPING ||
+      deploymentStatus === DeploymentStatus.STOPPED
+    ) {
+      setLastDeploymentStatus(DeploymentStatus.STOPPING)
+    } else if (deploymentStatus === DeploymentStatus.DEPLOYING) {
+      setLastDeploymentStatus(DeploymentStatus.DEPLOYING)
+    }
```
Comment (KwilLuke) [inline hooks/firebird/use-event-stream.tsx:131]: Still need to discuss with you the role of `lastDeploymentStatus`... Reading this code, it is kind of confusing. Why do you need it?

Assuming you do need `lastDeploymentStatus`, why are you checking if `deploymentStatus` is equal to some variable and then setting it to that variable if true?

Why don't you just setLastDeploymentStatus to whatever deployment status is??

`setLastDeploymentStatus(deploymentStatus)`

File: hooks/firebird/use-event-stream.tsx [hunk 4]
```diff
@@ -105,14 +110,47 @@ const useDeploymentEventStream = (deploymentId: string) => {
...
+    ) {
+      return "follow_deployment_stop"
+    } else if (deploymentStatus === DeploymentStatus.DEPLOYING) {
+      return "follow_deployment"
+    } else if (deploymentStatus === DeploymentStatus.ACTIVE) {
+      return lastDeploymentStatus === DeploymentStatus.STARTING
+        ? "follow_deployment_start"
+        : "follow_deployment"
+    }
+    return "follow_deployment"
```
Comment (KwilLuke) [inline hooks/firebird/use-event-stream.tsx:149]: Could you put this into something more readable like a switch statement? (Also, switch wouldn't require you to go through each case)

https://builtin.com/articles/switch-case-vs-if-else

File: components/Firebird/Deployments/ExistingDeployment/SelectedDeployment/StatusStream.tsx [hunk 5]
```diff
@@ -20,14 +20,24 @@ export default function DeploymentStatusStream({
...
-    [DeploymentEvents.INIT_KEY_PAIR]: "Initializing Key Pair",
-    [DeploymentEvents.CREATE_INSTANCE]: "Creating Instance",
-    [DeploymentEvents.WAIT_INSTANCE_READY]: "Starting Instance",
-    [DeploymentEvents.INSTALL_KWILD]: "Installing Kwil Daemon",
-    [DeploymentEvents.REGISTER_DOMAIN]: "Registering Domain",
-    [DeploymentEvents.FINALIZE_DEPLOYMENT]: "Finalizing Deployment",
-  }
+  // Display appropriate events based on event stream events
+  const eventDisplayNames =
+    status === DeploymentStatus.STARTING
```
Comment (KwilLuke) [inline components/Firebird/Deployments/ExistingDeployment/SelectedDeployment/StatusStream.tsx:25]: The nested ternary is quite hard to read.

Could you just create a utility function somewhere with a switch and then set it `eventDisplayNames` inside the functional component?

E.g.

```typescript
// utility function outside your functional component
const getEventDisplayNames = (status: DeploymentStatus) => {
  switch (status) {
    case DeploymentStatus.STARTING:
      return {
        [DeploymentEvents.START_INSTANCE]: "Starting Instance",
      };
    case DeploymentStatus.STOPPING:
      return {
        [DeploymentEvents.STOP_INSTANCE]: "Stopping Instance",
      };
    default:
      return {
        [DeploymentEvents.INIT_KEY_PAIR]: "Initializing Key Pair",
        [DeploymentEvents.CREATE_INSTANCE]: "Creating Instance",
        [DeploymentEvents.WAIT_INSTANCE_READY]: "Starting Instance",
        [DeploymentEvents.INSTALL_KWILD]: "Installing Kwil Daemon",
        [DeploymentEvents.REGISTER_DOMAIN]: "Registering Domain",
        [DeploymentEvents.FINALIZE_DEPLOYMENT]: "Finalizing Deployment",
      };
  }
};

// and then inside `StatusStream`: 
const eventDisplayNames = getEventDisplayNames(status);
```

**Review by KwilLuke — CHANGES_REQUESTED** [review]
Need to talk to you about a few things.

1. There is a slight issue in `useGetDeployment` where, when an instance switches from stopping to stopped, there is a brief moment where the node creation StatusStream appears.
2. I don't really understand the role of `lastDeploymentStatus` in `useEventStream`. Maybe we can talk about it live.

Also, I made some comments where I think readability should be improved.

**Comment by KwilLuke** [comment]
IDK why my review posted twice. Sorry about that 😆

File: components/Firebird/Deployments/ExistingDeployment/index.tsx [hunk 6]
```diff
@@ -9,12 +9,13 @@ import SelectedDeploymentTabs from "./SelectedDeploymentTabs"
...
 
 export default function ExistingDeployment({ id }: { id: string }) {
-  const deploymentEventStream = useDeploymentEventStream(id)
-  useGetDeployment(id, deploymentEventStream.status)
   const selectedDeployment = useAppSelector(selectSelectedDeployment)
   const isDeploymentActive =
     selectedDeployment?.status === DeploymentStatus.ACTIVE
 
+  const eventStream = useEventStream(id, selectedDeployment?.status)
+  useGetDeployment(id, eventStream.status)
```
[the remark this answers is not in the document]
Comment (tydangelo18) [reply components/Firebird/Deployments/ExistingDeployment/index.tsx:17]: @KwilLuke I understand why this makes sense, but, useGetDeployment is dependent on status

[the remark this answers is not in the document]
Comment (KwilLuke) [reply components/Firebird/Deployments/ExistingDeployment/index.tsx:17]: shit you're right

File: components/Firebird/Deployments/ExistingDeployment/SelectedDeployment/StatusStream.tsx [hunk 7]
```diff
@@ -20,13 +17,33 @@ export default function DeploymentStatusStream({
...
   if (status === DeploymentStatus.ACTIVE) return
 
-  const eventDisplayNames = {
-    [DeploymentEvents.INIT_KEY_PAIR]: "Initializing Key Pair",
-    [DeploymentEvents.CREATE_INSTANCE]: "Creating Instance",
-    [DeploymentEvents.WAIT_INSTANCE_READY]: "Starting Instance",
-    [DeploymentEvents.INSTALL_KWILD]: "Installing Kwil Daemon",
-    [DeploymentEvents.REGISTER_DOMAIN]: "Registering Domain",
-    [DeploymentEvents.FINALIZE_DEPLOYMENT]: "Finalizing Deployment",
+  let eventDisplayNames
```
Comment (KwilLuke) [inline components/Firebird/Deployments/ExistingDeployment/SelectedDeployment/StatusStream.tsx:20]: Shouldn't this be a useState?

**Review by KwilLuke — APPROVED** [review]
lgtm

**Pull request merged**